File: test-env/python-api/workload_sim.py
import math
import random
import threading
import time


class WorkloadSimulator:

    def __init__(
        self,
        max_cpu_ms=60,
        max_mem_mb=10,
        cycle_sec=2.5,
        personality="batch",
    ):
        self.max_cpu_ms = max_cpu_ms
        self.max_mem_mb = max_mem_mb
        self.cycle_sec = cycle_sec
        self.personality = personality
        self._mem_pool = []
        self._running = True
        self._start_time = time.time()

    def _intensity(self, elapsed):
        base = 0.5 + 0.3 * math.sin(2 * math.pi * elapsed / 300)

        if self.personality == "batch":
            burst_phase = elapsed % 60
            burst = 0.9 if burst_phase < 15 else 0.15
            base = base * 0.25 + burst * 0.75

        elif self.personality == "spike":
            roll = random.random()
            spike = 0.85 if roll < 0.08 else 0.25
            base = base * 0.35 + spike * 0.65

        elif self.personality == "business":
            sim_hour = (elapsed % 600) / 25
            if 8 <= sim_hour <= 18:
                day_factor = 0.4 + 0.5 * math.sin(
                    math.pi * (sim_hour - 8) / 10
                )
            else:
                day_factor = 0.12
            base = base * 0.15 + day_factor * 0.85

        base *= random.uniform(0.8, 1.2)
        return max(0.05, min(1.0, base))

    def _cpu_work(self, ms):
        end = time.time() + ms / 1000.0
        while time.time() < end:
            _ = math.sqrt(random.random()) * math.sin(random.random())

    def _adjust_memory(self, target_mb):
        current_mb = len(self._mem_pool)
        if target_mb > current_mb:
            alloc = min(target_mb - current_mb, 2)
            for _ in range(alloc):
                self._mem_pool.append(bytearray(1024 * 1024))
        elif target_mb < current_mb:
            release = max(1, (current_mb - target_mb) // 3)
            self._mem_pool = self._mem_pool[release:]

    def _touch_memory(self):
        if self._mem_pool:
            idx = random.randint(0, len(self._mem_pool) - 1)
            self._mem_pool[idx][random.randint(0, 1023)] = random.randint(0, 255)

    def _run(self):
        while self._running:
            elapsed = time.time() - self._start_time
            intensity = self._intensity(elapsed)

            cpu_ms = int(intensity * self.max_cpu_ms)
            self._cpu_work(cpu_ms)

            target_mb = max(1, int(intensity * self.max_mem_mb))
            self._adjust_memory(target_mb)
            self._touch_memory()

            sleep_sec = self.cycle_sec + random.uniform(-0.5, 0.5)
            time.sleep(max(0.5, sleep_sec))

    def start(self):
        t = threading.Thread(target=self._run, daemon=True)
        t.start()

    def stop(self):
        self._running = False

File: test-env/node-api/workload_sim.js
class WorkloadSimulator {
  constructor({
    maxCpuMs = 50,
    maxMemMb = 8,
    cycleSec = 2.5,
    personality = "spike",
  } = {}) {
    this.maxCpuMs = maxCpuMs;
    this.maxMemMb = maxMemMb;
    this.cycleSec = cycleSec;
    this.personality = personality;
    this.memPool = [];
    this.running = true;
    this.startTime = Date.now();
  }

  _intensity(elapsed) {
    const t = elapsed / 1000;
    let base = 0.5 + 0.3 * Math.sin((2 * Math.PI * t) / 300);

    if (this.personality === "spike") {
      const roll = Math.random();
      const spike = roll < 0.08 ? 0.85 : 0.25;
      base = base * 0.35 + spike * 0.65;
    } else if (this.personality === "batch") {
      const burstPhase = t % 60;
      const burst = burstPhase < 15 ? 0.9 : 0.15;
      base = base * 0.25 + burst * 0.75;
    } else if (this.personality === "business") {
      const simHour = (t % 600) / 25;
      let dayFactor;
      if (simHour >= 8 && simHour <= 18) {
        dayFactor = 0.4 + 0.5 * Math.sin((Math.PI * (simHour - 8)) / 10);
      } else {
        dayFactor = 0.12;
      }
      base = base * 0.15 + dayFactor * 0.85;
    }

    base *= 0.8 + Math.random() * 0.4;
    return Math.max(0.05, Math.min(1.0, base));
  }

  _cpuWork(ms) {
    const end = Date.now() + ms;
    while (Date.now() < end) {
      Math.sqrt(Math.random()) * Math.sin(Math.random());
    }
  }

  _adjustMemory(targetMb) {
    const currentMb = this.memPool.length;
    if (targetMb > currentMb) {
      const alloc = Math.min(targetMb - currentMb, 2);
      for (let i = 0; i < alloc; i++) {
        this.memPool.push(Buffer.alloc(1024 * 1024));
      }
    } else if (targetMb < currentMb) {
      const release = Math.max(1, Math.floor((currentMb - targetMb) / 3));
      this.memPool.splice(0, release);
    }
  }

  _touchMemory() {
    if (this.memPool.length > 0) {
      const idx = Math.floor(Math.random() * this.memPool.length);
      this.memPool[idx][Math.floor(Math.random() * 1024)] =
        Math.floor(Math.random() * 256);
    }
  }

  async _run() {
    while (this.running) {
      const elapsed = Date.now() - this.startTime;
      const intensity = this._intensity(elapsed);

      const cpuMs = Math.floor(intensity * this.maxCpuMs);
      this._cpuWork(cpuMs);

      const targetMb = Math.max(1, Math.floor(intensity * this.maxMemMb));
      this._adjustMemory(targetMb);
      this._touchMemory();

      const sleepSec = Math.max(
        0.5,
        this.cycleSec + (Math.random() - 0.5)
      );
      await new Promise((r) => setTimeout(r, sleepSec * 1000));
    }
  }

  start() {
    this._run().catch(console.error);
  }

  stop() {
    this.running = false;
  }
}

module.exports = { WorkloadSimulator };
